SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp

```diff
@@ -561,43 +561,57 @@ void USpatialSender::SendAuthorityIntentUpdate(const AActor& Actor, VirtualWorke
 	if (NetDriver->StaticComponentView->HasAuthority(EntityId, SpatialConstants::ENTITY_ACL_COMPONENT_ID))
 	{
 		// Also notify the enforcer directly on the worker that sends the component update, as the update will short circuit
-		NetDriver->LoadBalanceEnforcer->QueueAclAssignmentRequest(EntityId);
+		NetDriver->LoadBalanceEnforcer->MaybeQueueAclAssignmentRequest(EntityId);
 	}
 }
 
-void USpatialSender::SetAclWriteAuthority(const Worker_EntityId EntityId, const PhysicalWorkerName& DestinationWorkerId)
+void USpatialSender::SetAclWriteAuthority(
+	const Worker_EntityId EntityId,
+	const PhysicalWorkerName& DestinationWorkerId,
+	const WorkerRequirementSet& ReadAcl,
+	const WorkerRequirementSet& ClientRequirementSet,
+	const TArray<Worker_ComponentId>& ComponentIds)
 {
 	check(NetDriver);
-	check(NetDriver->StaticComponentView->HasAuthority(EntityId, SpatialConstants::ENTITY_ACL_COMPONENT_ID));
-
-	EntityAcl* EntityACL = NetDriver->StaticComponentView->GetComponentData<EntityAcl>(EntityId);
-	check(EntityACL);
 
 	const FString& WriteWorkerId = FString::Printf(TEXT("workerId:%s"), *DestinationWorkerId);
 
-	WorkerAttributeSet OwningWorkerAttribute = { WriteWorkerId };
+	const WorkerAttributeSet ACLAttributeSet = { SpatialConstants::ZoningAttribute };
+	const WorkerAttributeSet OwningServerWorkerAttributeSet = { WriteWorkerId };
+
+	EntityAcl NewAcl = EntityAcl();
 
-	TArray<Worker_ComponentId> ComponentIds;
-	EntityACL->ComponentWriteAcl.GetKeys(ComponentIds);
+	NewAcl.ReadAcl = ReadAcl;
 
 	for (const Worker_ComponentId& ComponentId : ComponentIds)
 	{
-		if (ComponentId == SpatialConstants::ENTITY_ACL_COMPONENT_ID ||
-			ComponentId == SpatialConstants::HEARTBEAT_COMPONENT_ID ||
-			ComponentId == SpatialConstants::GetClientAuthorityComponent(GetDefault<USpatialGDKSettings>()->UseRPCRingBuffer()))
+		if (ComponentId == SpatialConstants::HEARTBEAT_COMPONENT_ID
+			|| ComponentId == SpatialConstants::GetClientAuthorityComponent(GetDefault<USpatialGDKSettings>()->UseRPCRingBuffer()))
 		{
+			if (ClientRequirementSet.Num() == 1)
```

Comment:
Is there a valid case for having heartbeat/client endpoint IDs but not having a client requirement set?

Reply:
no, it looks like at least it will add a requirement set with an empty string for the requirement set regardless. Looks like to be consistent this should just pass through whatever the client requirement set is and assign it as is.

Reply:
changed